### PR TITLE
KubernetesApplicationResource to schedule to a cluster with only secret

### LINKED ISSUE #1605: Kubernetes Application cannot schedule to the types defined via InfrastructureDefinition
### What problem are you facing?

Kubernetes Application cannot schedule to the types defined via InfrastructureDefinition. This is required for [Wordpress](https://github.com/crossplane/app-wordpress) app to work with composition.

### How could Crossplane help solve your problem?

Allow `KubernetesApplication` or `KubernetesApplicationResource` to accept a secret to connect to the target cluster.

### PR DESCRIPTION
**Pull request by muvaf**
### Description of your changes

Claims are being deprecated and `KubernetesTarget` is tightly coupled to it. With composition, Crossplane will not know whether the type that is defined by `InfrastructureDefinition` has/is a Kubernetes cluster, hence, it could not schedule. The common denominator of all types that include a Kubernetes cluster is the credentials secret and this PR makes it possible to schedule to a target cluster by supplying a secret with kubeconfig.

With this PR, the actual connection is made with the `Secret` referenced in `spec.credentialsRef`. If only a `target` is given, then the controller extracts the secret reference from the target and writes it to `spec.credentialsRef` of KAR so that the change is backward-compatible.

Fixes #1605 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Manually with https://github.com/crossplane/app-wordpress/pull/68

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by muvaf** [comment]
Here is the latest update for the scenario that this PR wants to enable: https://github.com/crossplane/crossplane/issues/1586#issuecomment-648696616

@negz Making `KubernetesApplicationResource` work with bare `Secret`s will be beneficial for the point where we have new packages but not new app deployment model, so, I still think we should merge this one and make WP work with it using claim's secret.

File: apis/workload/v1alpha1/types.go [hunk 1]
```diff
@@ -170,12 +170,26 @@ type KubernetesApplicationResourceSpec struct {
 	// +optional
 	Target *KubernetesTargetReference `json:"targetRef,omitempty"`
 
+	// TargetCredentialsRef references the secret that has the credentials to
```
Comment (hasheddan) [inline apis/workload/v1alpha1/types.go:173]:
```suggestion
	// CredentialsRef references the secret that has the credentials to
```

**Comment by muvaf** [comment]
> Do you just mean for when a k8s cluster is provisioned using composition so it doesn't get bound and so no KubernetesTarget is automatically created for it?

Yeah, sort of. There will be a point where we still have to use `KubernetesApplication` but be blocked because of this. The agent stuff doesn't even have a design doc yet.

**Comment by muvaf** [comment]
What do you guys think about merging this? @negz @hasheddan

**Comment by muvaf** [comment]
@hasheddan The [issue](https://github.com/crossplane/crossplane/issues/1605) this PR fixes is closed. I will close this PR now and if we need it, we can re-open it.

**Pull request closed without merging**